### PR TITLE
[GCC13][AlCa] Avoid Wdangling-reference in SiPhase2OuterTrackerFakeLorentzAngleESSource

### PR DESCRIPTION
**Pull request by iarspider**
#### PR description:

In GCC13 IB, complier complains about dangling reference to a temporary:

```
src/CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc: In member function 'virtual std::unique_ptr<SiPhase2OuterTrackerLorentzAngle> SiPhase2OuterTrackerFakeLorentzAngleESSource::produceOTLA(const SiPhase2OuterTrackerLorentzAngleRcd&)':
  src/CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc:72:15: warning: possibly dangling reference to a temporary [-Wdangling-reference]
    72 |   const auto& geomDet = rcd.getRecord<TrackerTopologyRcd>().get(m_geomDetToken);
      |               ^~~~~~~
src/CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc:72:64: note: the temporary was destroyed at the end of the full expression 'edm::eventsetup::DependentRecordImplementation<RecordT, ListT>::getRecord() const [with DepRecordT = TrackerTopologyRcd; RecordT = SiPhase2OuterTrackerLorentzAngleRcd; ListT = edm::mpl::Vector<TrackerTopologyRcd>]().TrackerTopologyRcd::<anonymous>.edm::eventsetup::DependentRecordImplementation<TrackerTopologyRcd, edm::mpl::Vector<IdealGeometryRecord, PTrackerParametersRcd> >::get<GeometricDet, IdealGeometryRecord>(((SiPhase2OuterTrackerFakeLorentzAngleESSource*)this)->SiPhase2OuterTrackerFakeLorentzAngleESSource::m_geomDetToken)'
   72 |   const auto& geomDet = rcd.getRecord<TrackerTopologyRcd>().get(m_geomDetToken);
      |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
src/CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc: In member function 'virtual std::unique_ptr<SiPhase2OuterTrackerLorentzAngle> SiPhase2OuterTrackerFakeLorentzAngleESSource::produceOTSimLA(const SiPhase2OuterTrackerLorentzAngleSimRcd&)':
  src/CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc:78:15: warning: possibly dangling reference to a temporary [-Wdangling-reference]
    78 |   const auto& geomDet = rcd.getRecord<TrackerTopologyRcd>().get(m_geomDetToken);
      |               ^~~~~~~
src/CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc:78:64: note: the temporary was destroyed at the end of the full expression 'edm::eventsetup::DependentRecordImplementation<RecordT, ListT>::getRecord() const [with DepRecordT = TrackerTopologyRcd; RecordT = SiPhase2OuterTrackerLorentzAngleSimRcd; ListT = edm::mpl::Vector<TrackerTopologyRcd>]().TrackerTopologyRcd::<anonymous>.edm::eventsetup::DependentRecordImplementation<TrackerTopologyRcd, edm::mpl::Vector<IdealGeometryRecord, PTrackerParametersRcd> >::get<GeometricDet, IdealGeometryRecord>(((SiPhase2OuterTrackerFakeLorentzAngleESSource*)this)->SiPhase2OuterTrackerFakeLorentzAngleESSource::m_geomDetToken)'
   78 |   const auto& geomDet = rcd.getRecord<TrackerTopologyRcd>().get(m_geomDetToken);
      |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
```

Previously this warning was fixed by removing reference. However I think `geomDet` is not cheap to copy, so I opted for explicitly storing result of `getRecord` call.

#### PR validation:

Bot tests

### CODE REVIEW

**Comment by iarspider** [comment]
please test for el9_amd64_gcc13

**Comment by iarspider** [comment]
please test

**Comment by perrotta** [comment]
+alca

**Comment by srimanob** [comment]
+Upgrade

**Comment by makortel** [comment]
> Doesn't work.

Could you elaborate how it doesn't work?

**Comment by iarspider** [comment]
@makortel `error: static assertion failed: Trying to get a product with an ESGetToken specifying a Record from another Record where the second Record is not dependent on the first Record.`

**Comment by makortel** [comment]
Thanks. So the `m_geomDetToken` is ` edm::ESGetToken<GeometricDet, IdealGeometryRecord>`. Both `SiPhase2OuterTrackerLorentzAngleRcd` and `SiPhase2OuterTrackerLorentzAngleSimRcd` depend only on `TrackerTopologyRcd`, and `TrackerTopologyRcd` depends on `IdealGeometryRecord`.

@Dr15Jones Did we not make `DependentRecordImplementation::get{,Handle,TransientHandle}()` to look beyond the direct dependencies? On a quick look I'd be tempted to conclude that.

If this is the case, then in contrary to my earlier statement, there are some cases where the `getRecord()` intermediate step would still be needed.

**Comment by Dr15Jones** [comment]
> Did we not make DependentRecordImplementation::get{,Handle,TransientHandle}() to look beyond the direct dependencies? On a quick look I'd be tempted to conclude that.

If one wants the data from the other record, then one IS directly dependent upon that other record and it should be in the dependency list.

**Comment by mandrenguyen** [comment]
assign core

File: CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc [hunk 1]
```diff
@@ -69,13 +69,15 @@ namespace fakeOTLA {
 
 std::unique_ptr<SiPhase2OuterTrackerLorentzAngle> SiPhase2OuterTrackerFakeLorentzAngleESSource::produceOTLA(
     const SiPhase2OuterTrackerLorentzAngleRcd& rcd) {
-  const auto& geomDet = rcd.getRecord<TrackerTopologyRcd>().get(m_geomDetToken);
+  const auto& geomDetRcd = rcd.getRecord<TrackerTopologyRcd>();
+  const auto& geomDet = geomDetRcd.get(m_geomDetToken);
```
Comment (makortel) [inline CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc:73]: So the best course of action would be to
```suggestion
  const auto& geomDet = rcd.get(m_geomDetToken);
```
and to add `IdealGeometryRecord` is direct dependence of `SiPhase2OuterTrackerLorentzAngleRcd` and `SiPhase2OuterTrackerLorentzAngleSimRcd` in
https://github.com/cms-sw/cmssw/blob/3914bc83b3f9a0e3613f34c94bed4ec84f4b4cfe/CondFormats/DataRecord/interface/SiPhase2OuterTrackerCondDataRecords.h#L10-L15

**Comment by mmusich** [comment]
test parameters:
   * workflow = https://github.com/cms-sw/cmssw/pull/45810

**Comment by iarspider** [comment]
please test for el9_amd64_gcc13

**Comment by iarspider** [comment]
please test for CMSSW_14_1_X/el9_amd64_gcc13

File: CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc (hunk 1 above)
Comment (mmusich) [reply CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc:73]: @iarspider I think this part of the suggestion was not applied:

```diff
-  const auto& geomDetRcd = rcd.getRecord<TrackerTopologyRcd>();
-  const auto& geomDet = geomDetRcd.get(m_geomDetToken);
+  const auto& geomDet = rcd.get(m_geomDetToken);
```

File: CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc [hunk 2]
```diff
@@ -69,13 +69,14 @@ namespace fakeOTLA {
...
-  const auto& geomDet = rcd.getRecord<TrackerTopologyRcd>().get(m_geomDetToken);
+  const auto& geomDet = rcd.get(m_geomDetToken);
   return fakeOTLA::produceRecord<SiPhase2OuterTrackerLorentzAngle>(LAvalue_, geomDet);
 }
 
 std::unique_ptr<SiPhase2OuterTrackerLorentzAngle> SiPhase2OuterTrackerFakeLorentzAngleESSource::produceOTSimLA(
     const SiPhase2OuterTrackerLorentzAngleSimRcd& rcd) {
-  const auto& geomDet = rcd.getRecord<TrackerTopologyRcd>().get(m_geomDetToken);
+  const auto& geomDetRcd = rcd.getRecord<TrackerTopologyRcd>();
+  const auto& geomDet = geomDetRcd.get(m_geomDetToken);
```
Comment (mmusich) [inline CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2OuterTrackerFakeLorentzAngleESSource.cc:79]:
```suggestion
  const auto& geomDet = rcd.get(m_geomDetToken);
```

here as well.

**Comment by mmusich** [comment]
test parameters:

**Comment by mmusich** [comment]
please test for el9_amd64_gcc13

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1

**Comment by srimanob** [comment]
+Upgrade

Technical PR.

**Pull request merged**